## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [325]:
import pandas as pd
import sys
# sys.path.append('shipping_calculator.py')
# from shipping_calculator import calculate_shipping_price


client_df = pd.read_csv('client_dataset.csv')
client_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [326]:
# View the column names in the data
client_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [327]:
# Use the describe function to gather some basic statistics
client_df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [328]:
# Use this space to do any additional research
# and familiarize yourself with the data.
client_df['client_name'] = client_df['first'] + " " + client_df['last']
client_df = client_df[['client_id', 'client_name', 'order_date' , 'order_id', 'category' , 'subcategory', 'unit_price', 'unit_cost' , 'qty', 'unit_weight', 'line_number']]
client_df.head()

client_id         client_name  order_date  order_id     category  \
0      58515      Donald Harding  2023-04-28   8953482        decor   
1      37609       Tiffany Myers  2023-05-19   8069089  consumables   
2      57113      Shannon Watson  2023-01-29   1902144     software   
3      46554        Nathan Baker  2023-04-25   9031802  consumables   
4      92089  Christina Schwartz  2023-05-28   1322274  consumables   

          subcategory  unit_price  unit_cost  qty  unit_weight  line_number  
0            wall art     1096.80     762.71  105         7.50            1  
1                pens       24.95      15.09   21         1.49            0  
2  project management       13.52       7.86   39         1.68            6  
3                pens       36.42      24.85   29         1.23            3  
4                misc      195.10     108.17   20        46.43            1

In [329]:
# What three item categories had the most entries?

# Gets the top three caegories with the most entries
top_categories = client_df['category'].value_counts().head(3)

top_categories.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [330]:
# For the category with the most entries, which subcategory had the most entries?
client_df.loc[client_df.category == 'consumables' , 'subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [331]:
# Which five clients had the most entries in the data?
top_clients_list = client_df['client_name'].value_counts()
top_clients_list.head(5)



client_name
Jessica Reyes      220
Angela Everett     211
Bryan Myers        209
Alexandra Young    207
Kendra Garrett     207
Name: count, dtype: int64

In [332]:
top_clients = client_df['client_id'].value_counts().head(5)

# Store the client ids of those top 5 clients in a list.
top_client_ids = top_clients.index.tolist()

# Print the list to the top 5 client ids
print(f"Top 5 client IDs:\n", top_client_ids)

Top 5 client IDs:
 [33615, 66037, 46820, 38378, 24741]


In [333]:
# How many total units (the qty column) did the client with the most entries order?

# Find the customer with the most entries
top_client = client_df['client_name'].value_counts().head().idxmax()

# Filter for the top customer
top_client_data = client_df[client_df['client_name'] == top_client]

# Calculate the total quantity ordered by the top cusstomer
top_client_units = top_client_data['qty'].sum()

# Print the results
print(f"The total units ordered by {top_client} was: {top_client_units}")


The total units ordered by Jessica Reyes was: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [334]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
client_df.loc[:,'line_subtotal'] = client_df['unit_price'] * client_df['qty']

# Display the updated DataFrame
client_df.head()

client_id         client_name  order_date  order_id     category  \
0      58515      Donald Harding  2023-04-28   8953482        decor   
1      37609       Tiffany Myers  2023-05-19   8069089  consumables   
2      57113      Shannon Watson  2023-01-29   1902144     software   
3      46554        Nathan Baker  2023-04-25   9031802  consumables   
4      92089  Christina Schwartz  2023-05-28   1322274  consumables   

          subcategory  unit_price  unit_cost  qty  unit_weight  line_number  \
0            wall art     1096.80     762.71  105         7.50            1   
1                pens       24.95      15.09   21         1.49            0   
2  project management       13.52       7.86   39         1.68            6   
3                pens       36.42      24.85   29         1.23            3   
4                misc      195.10     108.17   20        46.43            1   

   line_subtotal  
0      115164.00  
1         523.95  
2         527.28  
3        1056.18  
4        3902.00

In [335]:
sys.path.append('shipping_calculator.py')
from shipping_calculator import calculate_shipping_price

# Create a column for shipping price.
# Assume a shipping price of $7 per pound for items over 50 pounds and $10 per pound for items 50 pounds or under.

client_df['shipping_price']  = client_df.apply(
    lambda row: calculate_shipping_price(row['unit_weight'], row['qty']), axis=1
)

client_df.head()

client_id         client_name  order_date  order_id     category  \
0      58515      Donald Harding  2023-04-28   8953482        decor   
1      37609       Tiffany Myers  2023-05-19   8069089  consumables   
2      57113      Shannon Watson  2023-01-29   1902144     software   
3      46554        Nathan Baker  2023-04-25   9031802  consumables   
4      92089  Christina Schwartz  2023-05-28   1322274  consumables   

          subcategory  unit_price  unit_cost  qty  unit_weight  line_number  \
0            wall art     1096.80     762.71  105         7.50            1   
1                pens       24.95      15.09   21         1.49            0   
2  project management       13.52       7.86   39         1.68            6   
3                pens       36.42      24.85   29         1.23            3   
4                misc      195.10     108.17   20        46.43            1   

   line_subtotal  shipping_price  
0      115164.00         5512.50  
1         523.95          312.90  
2         527.28          458.64  
3        1056.18          356.70  
4        3902.00         6500.20

In [336]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

# Calculate the total price
client_df['total_price'] = client_df.apply(
    lambda x: (x['line_subtotal'] + x['shipping_price']) *1.0925, axis=1).round(2)

# Display the updated data
client_df.head()

client_id         client_name  order_date  order_id     category  \
0      58515      Donald Harding  2023-04-28   8953482        decor   
1      37609       Tiffany Myers  2023-05-19   8069089  consumables   
2      57113      Shannon Watson  2023-01-29   1902144     software   
3      46554        Nathan Baker  2023-04-25   9031802  consumables   
4      92089  Christina Schwartz  2023-05-28   1322274  consumables   

          subcategory  unit_price  unit_cost  qty  unit_weight  line_number  \
0            wall art     1096.80     762.71  105         7.50            1   
1                pens       24.95      15.09   21         1.49            0   
2  project management       13.52       7.86   39         1.68            6   
3                pens       36.42      24.85   29         1.23            3   
4                misc      195.10     108.17   20        46.43            1   

   line_subtotal  shipping_price  total_price  
0      115164.00         5512.50    131839.08  
1         523.95          312.90       914.26  
2         527.28          458.64      1077.12  
3        1056.18          356.70      1543.57  
4        3902.00         6500.20     11364.40

In [337]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

client_df['line_cost'] = (client_df['unit_cost'] * client_df['qty']) + client_df['shipping_price']

client_df.head()



client_id         client_name  order_date  order_id     category  \
0      58515      Donald Harding  2023-04-28   8953482        decor   
1      37609       Tiffany Myers  2023-05-19   8069089  consumables   
2      57113      Shannon Watson  2023-01-29   1902144     software   
3      46554        Nathan Baker  2023-04-25   9031802  consumables   
4      92089  Christina Schwartz  2023-05-28   1322274  consumables   

          subcategory  unit_price  unit_cost  qty  unit_weight  line_number  \
0            wall art     1096.80     762.71  105         7.50            1   
1                pens       24.95      15.09   21         1.49            0   
2  project management       13.52       7.86   39         1.68            6   
3                pens       36.42      24.85   29         1.23            3   
4                misc      195.10     108.17   20        46.43            1   

   line_subtotal  shipping_price  total_price  line_cost  
0      115164.00         5512.50    131839.08   85597.05  
1         523.95          312.90       914.26     629.79  
2         527.28          458.64      1077.12     765.18  
3        1056.18          356.70      1543.57    1077.35  
4        3902.00         6500.20     11364.40    8663.60

In [338]:
# Create a column for the profit of each line using line cost and line price

client_df['line_profit'] = client_df['total_price'] - client_df['line_cost']

client_df.head()

client_id         client_name  order_date  order_id     category  \
0      58515      Donald Harding  2023-04-28   8953482        decor   
1      37609       Tiffany Myers  2023-05-19   8069089  consumables   
2      57113      Shannon Watson  2023-01-29   1902144     software   
3      46554        Nathan Baker  2023-04-25   9031802  consumables   
4      92089  Christina Schwartz  2023-05-28   1322274  consumables   

          subcategory  unit_price  unit_cost  qty  unit_weight  line_number  \
0            wall art     1096.80     762.71  105         7.50            1   
1                pens       24.95      15.09   21         1.49            0   
2  project management       13.52       7.86   39         1.68            6   
3                pens       36.42      24.85   29         1.23            3   
4                misc      195.10     108.17   20        46.43            1   

   line_subtotal  shipping_price  total_price  line_cost  line_profit  
0      115164.00         5512.50    131839.08   85597.05     46242.03  
1         523.95          312.90       914.26     629.79       284.47  
2         527.28          458.64      1077.12     765.18       311.94  
3        1056.18          356.70      1543.57    1077.35       466.22  
4        3902.00         6500.20     11364.40    8663.60      2700.80

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [339]:
# Check your work using the totals above

order_totals = client_df.groupby('order_id')[['total_price']].sum()
order_totals['line_total_cost'] = order_totals['total_price']

# List of order_ids to filter
specific_order_ids = [2742071, 2173913, 6128929]

# Filter the grouped DataFrame for the specific order IDs
specific_orders = order_totals[order_totals.index.isin(specific_order_ids)]

# Display the result
print(specific_orders)

# Verify the receipts
expected_prices = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

          total_price  line_total_cost
order_id                              
2173913     162388.71        162388.71
2742071     152811.89        152811.89
6128929     923441.25        923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [340]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

# Function to sum the total price based on client_id
def client_totals(client_id, column):
    df = client_df.loc[client_df['client_id']== client_id , column]
    return round(df.sum(),2)

# Call function and display results
for client_id in top_client_ids:
    print(f"{client_id}: ${client_totals(client_id, 'total_price')}")

33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [341]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.

top_clients_df = client_df[client_df['client_id'].isin(top_client_ids)]

# Group by 'client_id' and sum
summary = top_clients_df.groupby('client_id').agg(
    Client_Name=('client_name', 'first'),
    Total_Units_Purchased=('qty', 'sum'),
    Total_Shipping_Price=('shipping_price', 'sum'),
    Total_Revenue=('total_price', 'sum'),
    Total_Profit=('line_profit', 'sum')
).reset_index()

# Sort the summary DataFrame by Total_Revenue in descending order
summary = summary.sort_values(by='Total_Revenue', ascending=False)

# Display the sorted summary DataFrame
print("Summary DataFrame sorted by Total Revenue (descending):")
summary

Summary DataFrame sorted by Total Revenue (descending):


client_id      Client_Name  Total_Units_Purchased  Total_Shipping_Price  \
0      24741   Kendra Garrett                 239862            5126448.37   
2      38378  Alexandra Young                  73667            3429455.40   
4      66037   Angela Everett                  43018            1395151.85   
3      46820      Bryan Myers                  75768            1601448.84   
1      33615    Jessica Reyes                  64313            1828984.89   

   Total_Revenue  Total_Profit  
0    82268892.02   36579992.31  
2    12906550.87    3271829.89  
4    10259514.79    3255031.81  
3     9743794.36    2736602.72  
1     8377308.52    2201994.61

In [342]:
# Format the data and rename the columns to names suitable for presentation.
columns_to_convert = ['Total_Shipping_Price', 'Total_Revenue', 'Total_Profit']
summary[columns_to_convert] = summary[columns_to_convert].apply(pd.to_numeric, errors='coerce')

# Handle NaN values
summary[columns_to_convert] = summary[columns_to_convert].fillna(0)
# Define the money columns. 
money_columns =['Total_Shipping_Price', 'Total_Revenue', 'Total_Profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    # return f"${x / 1_000_000:.2f}" if pd.notnull(x) else "$0.00"
    return x / 1000000
# Apply the currency_format_millions function to only the money columns. 
summary[money_columns] = summary[money_columns].map(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
summary.rename(columns={
    'client_id': 'Client ID',
    'Total_Shipping_Price': 'Total Shipping Price (in millions)',
    'Total_Revenue': 'Total Revenue (in millions)',
    'Total_Profit': 'Total Profit (in millions)'
               }, inplace=True)

print(summary)

   Client ID      Client_Name  Total_Units_Purchased  \
0      24741   Kendra Garrett                 239862   
2      38378  Alexandra Young                  73667   
4      66037   Angela Everett                  43018   
3      46820      Bryan Myers                  75768   
1      33615    Jessica Reyes                  64313   

   Total Shipping Price (in millions)  Total Revenue (in millions)  \
0                            5.126448                    82.268892   
2                            3.429455                    12.906551   
4                            1.395152                    10.259515   
3                            1.601449                     9.743794   
1                            1.828985                     8.377309   

   Total Profit (in millions)  
0                   36.579992  
2                    3.271830  
4                    3.255032  
3                    2.736603  
1                    2.201995  


In [343]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
# Sort the summary DataFrame by Total_Revenue in descending order
summary_by_total_revenue = summary.sort_values(by='Total Revenue (in millions)', ascending=False)

# Display the sorted summary DataFrame
print("Summary DataFrame sorted by Total Revenue (descending):")
summary_by_total_revenue


Summary DataFrame sorted by Total Revenue (descending):


Client ID      Client_Name  Total_Units_Purchased  \
0      24741   Kendra Garrett                 239862   
2      38378  Alexandra Young                  73667   
4      66037   Angela Everett                  43018   
3      46820      Bryan Myers                  75768   
1      33615    Jessica Reyes                  64313   

   Total Shipping Price (in millions)  Total Revenue (in millions)  \
0                            5.126448                    82.268892   
2                            3.429455                    12.906551   
4                            1.395152                    10.259515   
3                            1.601449                     9.743794   
1                            1.828985                     8.377309   

   Total Profit (in millions)  
0                   36.579992  
2                    3.271830  
4                    3.255032  
3                    2.736603  
1                    2.201995